File: 851.ts
function loudAndRich(richer: number[][], quiet: number[]): number[] {
  const graph = Array.from({ length: quiet.length }, (): number[] => []);
  for (const [u, v] of richer) {
    graph[v].push(u);
  }

  const ans = new Array<number>(quiet.length).fill(-1);

  const dfs = (v: number): number => {
    if (ans[v] !== -1) {
      return ans[v];
    }
    let richestNode = v;

    for (const u of graph[v]) {
      const richerNode = dfs(u);
      if (quiet[richestNode] > quiet[richerNode]) {
        richestNode = richerNode;
      }
      //console.log(v, richerNode, richestNode);
    }

    ans[v] = richestNode;
    return richestNode;
  };

  for (let i = 0; i < quiet.length; ++i) {
    dfs(i);
  }

  return ans;
}

File: 851.cpp
class Solution {
public:
    int DFS(int u, std::vector<std::vector<int>>& adjacency, std::vector<bool>& visited, std::vector<int>& answer, std::vector<int>& quiet) {
        int currentMin = quiet[u];
        int currentMinNode = u;
        visited[u] = true;
        for (auto v : adjacency[u]) {
            if (visited[v]) {
                if (currentMin > quiet[answer[v]]) {
                    currentMin = quiet[answer[v]];
                    currentMinNode = answer[v];
                }
            }
            else {
                int minNodeV = DFS(v, adjacency, visited, answer, quiet);
                if (currentMin > quiet[minNodeV]) {
                    currentMin = quiet[minNodeV];
                    currentMinNode = minNodeV;
                }
            }
        }
        answer[u] = currentMinNode;
        return currentMinNode;
    }

    std::vector<int> loudAndRich(std::vector<std::vector<int>>& richer, std::vector<int>& quiet) {
        int n = quiet.size();
        std::vector<int> answer(n);
        std::vector<std::vector<int>> adjacency(n);
        for (int i = 0; i < richer.size(); ++i) {
            adjacency[richer[i][1]].push_back(richer[i][0]); // reverse edges
        }
        std::vector<bool> visited(n, false);
      // using topsort idea
        for (int u = 0; u < n; ++u) {
            if (!visited[u]) {
                DFS(u, adjacency, visited, answer, quiet);
            }
        }
        return answer;
    }
};
